### ISSUE TITLE
UFRM crashes when soil data has a nodata of zero

### ISSUE THREAD

**Issue by dcdenu4** [natcap/invest#185]
This issue was brought to attention via two users on the forums:
https://community.naturalcapitalproject.org/t/urban-flood-risk-mitigation-soil-group-raster-preparation-error/1066/6
https://community.naturalcapitalproject.org/t/urban-flood-mitigation-model-run-error/918/8

The error being returned was:
```
Traceback (most recent call last):
File “site-packages\natcap\invest\ui\execution.py”, line 68, in run
File “site-packages\natcap\invest\ui\model.py”, line 1632, in _logged_target
File “site-packages\natcap\invest\urban_flood_risk_mitigation.py”, line 248, in execute
File “site-packages\taskgraph\Task.py”, line 592, in add_task
File “site-packages\taskgraph\Task.py”, line 1069, in _call
File “site-packages\pygeoprocessing\geoprocessing.py”, line 424, in raster_calculator
File “site-packages\natcap\invest\urban_flood_risk_mitigation.py”, line 832, in _lu_to_cn_op
ValueError: invalid entry in choice array
Check that the Soil Group raster does not contain values other than (1, 2, 3, 4)
2020-06-15 11:53:35,063 execution.run(83) INFO Execution finished
```

After lots of debugging it appears to me that soil group raster having a nodata value of 0 was the cause. This ends up being an issue in UFRM `_lu_to_cn_op`. I still need to really isolate why this is failing as the masking order seems okay at first glance.

Along the way I noticed something else when testing the following code from UFRM:
```python
# make cn_table into a 2d array where first dim is lucode, second is
# 0..3 to correspond to CN_A..CN_D
data = []
row_ind = []
col_ind = []
for lucode in cn_table:
    data.extend([
        cn_table[lucode]['cn_%s' % soil_id]
        for soil_id in ['a', 'b', 'c', 'd']])
    row_ind.extend([int(lucode)] * 4)
col_ind = [0, 1, 2, 3] * (len(row_ind) // 4)
lucode_to_cn_table = scipy.sparse.csr_matrix((data, (row_ind, col_ind)))
```
The `row_ind` produced here will start from `1` and not from `0`. This, in my testing, causes a row of `0`s in `lucode_to_cn_table`. Here is some testing code:
```python
bio_table = {1: {'lucode': 1, 'cn_a': 89, 'cn_b': 92, 'cn_c': 94, 'cn_d': 95}, 2: {'lucode': 2, 'cn_a': 49, 'cn_b': 69, 'cn_c': 79, 'cn_d': 84}, 3: {'lucode': 3, 'cn_a': 81, 'cn_b': 88, 'cn_c': 91, 'cn_d': 93}, 4: {'lucode': 4, 'cn_a': 89, 'cn_b': 92, 'cn_c': 94, 'cn_d': 95}, 5: {'lucode': 5, 'cn_a': 89, 'cn_b': 92, 'cn_c': 94, 'cn_d': 95}}
data = []
row_ind = []
col_ind = []
for lucode in bio_table:
     data.extend([bio_table[lucode]['cn_%s' % soil_id] for soil_id in ['a', 'b', 'c', 'd']])
     row_ind.extend([int(lucode)] * 4)
col_ind = [0, 1, 2, 3] * (len(row_ind) // 4)

lucode_to_cn_table = scipy.sparse.csr_matrix((data, (row_ind, col_ind)))
# This is python console output:
lucode_to_cn_table.toarray()
array([[ 0,  0,  0,  0],
       [89, 92, 94, 95],
       [49, 69, 79, 84],
       [81, 88, 91, 93],
       [89, 92, 94, 95],
       [89, 92, 94, 95]], dtype=int32)

row_ind
[1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5]
# END python console output
```
It is unclear to me if this is desired or a bug also.

**Comment by dcdenu4** [comment]
@phargogh I was going to take a stab at this and have you review it in the PR. If you'd like to take ownership, feel free! I'm currently working on this in the branch ` bugfix/ufrm-nodata-zero-bug` off of my fork.

**Comment by richpsharp** [comment]
Hey @dcdenu4, I think I implemented this model back in the day. I don't have the data so I'm gonna have to guess here that they have a 0 in their raster but it's not actually set as the nodata value. If that's not the case then yes, that's a bug. Otherwise I put an explicit check in there that the soil IDs were exactly 1, 2, 3, or 4 because the sample data we got had a 0 and a 5 in it which corresponded to another soil classification scheme. So that error check there was to guard against that. 

Offhand I'd say the user is providing bad data by providing a raster that does not have the nodata value set. I can't tell if one user makes this a common issue, but if so, I'd replace that exception to a warning and then only process values 1-4. If the user provides an incorrect soil ID raster like the case I described above, they'd have weird output.

So either case here you're dealing with bad user input and it's just a question of how you'd want to handle it.

**Comment by richpsharp** [comment]
Oop I take that back @dcdenu4, the soil raster looks well formed. I'm guessing there's another error but I can't tell what it is because the user only posted the soil raster. Do you have their entire data stack? And if so can you pass it here? I'd be happy to debug. Otherwise I wouldn't make changes to the model unless we can recreate that error.

**Comment by richpsharp** [comment]
I guess in summary I'm saying that I do not believe this bug is caused by a nodata value of 0 on the soil raster. I think it's something else.

**Comment by phargogh** [comment]
@dcdenu4 feel free to take this!  Very happy to review if you like.

**Comment by dcdenu4** [comment]
@richpsharp, I'm just finishing patching this now! It was indeed an issue in the `_lu_to_cn_op` where:
```python
if soil_type_nodata:
    valid_mask[:] &= ~numpy.isclose(soil_type_array, soil_type_nodata)
```
was being used. So if nodata here was 0 then it wouldn't be properly masked, leaving 0s in the array that would then cause the error.

Hopefully this makes sense!

**Comment by dcdenu4** [comment]
I'll put you on the PR as well!

**Comment by dcdenu4** [comment]
It took me some time to understand the usage of the sparse matrix but I don't think that is an issue and is behaving like it should. Very neat use of it.

**Comment by richpsharp** [comment]
Good catch @dcdenu4! Thanks for putting me on the review. I just did my review. :)

**Comment by dcdenu4** [comment]
resolved in #231
